File: py-methods/filtering_performance.py
import os
import time
import argparse
import scanpy as sc
import pandas as pd
import numpy as np
import gc

def log_benchmark_result(file_handle, test_case, step, duration_ms, success, dimensions=None, filtered_count=None):
    """Log a benchmark result to the CSV file."""
    dims_str = f"{dimensions[0]},{dimensions[1]}" if dimensions else ","
    filtered_str = str(filtered_count) if filtered_count is not None else ""
    
    file_handle.write(f"{test_case},{step},{duration_ms},{success},{dims_str},{filtered_str}\n")

def run_filtering_benchmark(h5ad_path, csv_path, iterations):
    """Run the filtering benchmark on an h5ad file."""
    print("Starting filtering benchmark")
    
    # Create or open the output CSV file
    file_exists = os.path.isfile(csv_path)
    with open(csv_path, 'a') as file:
        # Add header if file is new
        if not file_exists:
            file.write("test_case,test_step,time_ms,success,rows,cols,filtered_count\n")
        
    
        if True:
            print(f"Running filtering benchmarks for {iterations} iterations")
            
            # Combined filtering benchmark
            for i in range(1, iterations + 1):
                print(f"  Combined filtering - Iteration {i}/{iterations}")
                
                adata_copy = sc.read_h5ad(h5ad_path)
                n_obs = adata_copy.n_obs
                n_vars = adata_copy.n_vars
                
                combined_start = time.time()
                try:
                    # Calculate QC metrics if they don't exist
                    if 'n_genes_by_counts' not in adata_copy.obs or 'n_cells_by_counts' not in adata_copy.var:
                        sc.pp.calculate_qc_metrics(adata_copy, inplace=True)
                    
                    initial_n_obs = adata_copy.n_obs
                    initial_n_vars = adata_copy.n_vars
                    
                    # Cell filtering with scanpy functions
                    sc.pp.filter_cells(adata_copy, min_genes=200)
                    sc.pp.filter_cells(adata_copy, max_genes=5000)
                    sc.pp.filter_cells(adata_copy, min_counts=1000)
                    
                    # Gene filtering with scanpy functions
                    sc.pp.filter_genes(adata_copy, min_cells=3)
                    sc.pp.filter_genes(adata_copy, min_counts=10)
                    max_cells = int(0.9 * adata_copy.n_obs)
                    sc.pp.filter_genes(adata_copy, max_cells=max_cells)
                    
                    # Calculate filtered counts
                    filtered_cells = initial_n_obs - adata_copy.n_obs
                    filtered_genes = initial_n_vars - adata_copy.n_vars
                    filtered_total = filtered_cells + filtered_genes
                    combined_success = True
                except Exception as e:
                    print(f"Combined filtering failed: {e}")
                    filtered_total = None
                    combined_success = False
                
                del adata_copy
                gc.collect()
                
                combined_time = int((time.time() - combined_start) * 1000)
                
                
                log_benchmark_result(
                    file,
                    "combined_filtering",
                    "filter",
                    combined_time,
                    combined_success,
                    dimensions=(n_obs, n_vars),
                    filtered_count=filtered_total
                )
            
            print("Filtering benchmarks completed successfully")

def main():
    # Parse command line arguments
    parser = argparse.ArgumentParser(
        description="Benchmark for H5AD filtering operations",
    )
    parser.add_argument('-i', '--input', dest='h5ad_path', required=True,
                        help='Path to H5AD file')
    parser.add_argument('-o', '--output', dest='csv_path', required=True,
                        help='Path for output CSV file')
    parser.add_argument('-n', '--iterations', dest='iterations', type=int, default=5,
                        help='Number of benchmark iterations to run')
    
    args = parser.parse_args()
    
    print("Starting filtering benchmark with:")
    print(f"  Input file: {args.h5ad_path}")
    print(f"  Output file: {args.csv_path}")
    print(f"  Iterations: {args.iterations}")
    
    run_filtering_benchmark(args.h5ad_path, args.csv_path, args.iterations)
    
    print("Benchmark completed")

if __name__ == "__main__":
    main()

File: rs-methods/src/filtering_performance.rs
use anndata::AnnDataOp;
use anyhow::Result;
use clap::Parser;
use ndarray::Array1;
use single_rust::io;
use single_rust::memory::processing::filtering::{mark_filter_cells, mark_filter_genes};
use std::fs::{File, OpenOptions};
use std::io::Write;
use std::path::PathBuf;
use std::time::Instant;

#[derive(Parser)]
#[command(
    author = "Ian Ferenc Diks",
    version,
    about = "Benchmark for H5AD filtering operations"
)]
struct Args {
    /// Path to H5AD file
    #[arg(short = 'i', long = "input")]
    h5ad_path: PathBuf,
    /// Path for output CSV file
    #[arg(short = 'o', long = "output")]
    csv_path: PathBuf,
    /// Number of benchmark iterations to run
    #[arg(short = 'n', long = "iterations", default_value = "5")]
    iterations: usize,
}

fn log_benchmark_result(
    file: &mut File,
    test_case: &str,
    step: &str,
    duration_ms: u128,
    success: bool,
    dimensions: Option<(usize, usize)>,
    filtered_count: Option<usize>,
) -> Result<()> {
    let dims_str = if let Some((obs, vars)) = dimensions {
        format!("{},{}", obs, vars)
    } else {
        String::from(",")
    };

    let filtered_str = if let Some(count) = filtered_count {
        count.to_string()
    } else {
        String::from("")
    };

    writeln!(
        file,
        "{},{},{},{},{},{}",
        test_case, step, duration_ms, success, dims_str, filtered_str
    )?;
    Ok(())
}

fn run_filtering_benchmark(
    h5ad_path: &PathBuf,
    csv_path: &PathBuf,
    iterations: usize,
) -> Result<()> {
    println!("Starting filtering benchmark");

    // Create or open the output CSV file
    let mut file = OpenOptions::new()
        .create(true)
        .append(true)
        .open(csv_path)?;

    // Add header if file is new
    if file.metadata()?.len() == 0 {
        writeln!(
            file,
            "test_case,test_step,time_ms,success,rows,cols,filtered_count"
        )?;
    }

    // Initial load to get dimensions
    println!("Loading dataset once to get dimensions");
    let temp_load_start = Instant::now();
    let temp_adata = io::read_h5ad(h5ad_path, io::FileScope::Read, false);
    let temp_load_time = temp_load_start.elapsed().as_millis();

    let (n_obs, n_vars) = if let Ok(adata) = &temp_adata {
        (adata.n_obs(), adata.n_vars())
    } else {
        log_benchmark_result(
            &mut file,
            "initial_dimensions_check",
            "load",
            temp_load_time,
            false,
            None,
            None,
        )?;
        println!("Failed to load dataset for initial dimension check");
        return Err(anyhow::anyhow!(
            "Failed to load dataset for initial dimension check"
        ));
    };

    println!("Dataset dimensions: {} cells x {} genes", n_obs, n_vars);

    // Drop the temporary dataset
    drop(temp_adata);

    // Combined filtering benchmark
    for i in 1..=iterations {
        println!("  Combined filtering - Iteration {}/{}", i, iterations);

        // Load fresh data for this iteration
        let load_start = Instant::now();
        let adata_result = io::read_h5ad_fast_memory(h5ad_path);
        let load_time = load_start.elapsed().as_millis();

        if let Ok(mut adata) = adata_result {
            log_benchmark_result(
                &mut file,
                "combined_filtering",
                "load",
                load_time,
                true,
                Some((n_obs, n_vars)),
                None,
            )?;

            let combined_start = Instant::now();

            // First filter cells
            let cell_filter_result = mark_filter_cells::<u32, f64>(
                &adata,
                Some(200),    // min_genes
                Some(5000),   // max_genes
                Some(1000.0), // min_counts
                None,         // max_counts
                None,         // min_fraction
                None,         // max_fraction
            );

            // Then filter genes
            let gene_filter_result = if cell_filter_result.is_ok() {
                mark_filter_genes::<u32, f64>(
                    &adata,
                    Some(3),    // min_cells
                    None,       // max_cells
                    Some(10.0), // min_counts
                    None,       // max_counts
                    None,       // min_fraction
                    Some(0.9),  // max_fraction
                )
            } else {
                Err(anyhow::anyhow!("Cell filtering failed"))
            };

            // Calculate filtered counts if both operations succeeded
            if let (Ok(cell_filter), Ok(gene_filter)) = (cell_filter_result, gene_filter_result) {
                // Convert to Array1 and perform the subsetting
                let cell_filter = Array1::from(cell_filter);
                let gene_filter = Array1::from(gene_filter);

                let obs_sel_info =
                    single_rust::shared::get_select_info_obs(Some(cell_filter.view()))?;
                let var_sel_info =
                    single_rust::shared::get_select_info_vars(Some(gene_filter.view()))?;

                // Apply both filters
                adata.subset_inplace(&[&obs_sel_info[0], &var_sel_info[1]])?;

                let combined_time = combined_start.elapsed().as_millis();
                let filtered_total = (n_obs - adata.n_obs()) + (n_vars - adata.n_vars());

                log_benchmark_result(
                    &mut file,
                    "combined_filtering",
                    "filter",
                    combined_time,
                    true,
                    Some((n_obs, n_vars)),
                    Some(filtered_total),
                )?;
            } else {
                let combined_time = combined_start.elapsed().as_millis();
                log_benchmark_result(
                    &mut file,
                    "combined_filtering",
                    "filter",
                    combined_time,
                    false,
                    Some((n_obs, n_vars)),
                    None,
                )?;
            }
        } else {
            log_benchmark_result(
                &mut file,
                "combined_filtering",
                "load",
                load_time,
                false,
                Some((n_obs, n_vars)),
                None,
            )?;
            println!(
                "Failed to load dataset for combined filtering iteration {}",
                i
            );
        }
    }

    println!("Filtering benchmarks completed successfully");

    Ok(())
}

fn main() -> Result<()> {
    let args = Args::parse();
    println!("Starting filtering benchmark with:");
    println!("  Input file: {}", args.h5ad_path.display());
    println!("  Output file: {}", args.csv_path.display());
    println!("  Iterations: {}", args.iterations);

    run_filtering_benchmark(&args.h5ad_path, &args.csv_path, args.iterations)?;

    println!("Benchmark completed");
    Ok(())
}
